src/oidc_child/oidc_child_json.c

```diff
@@ -0,0 +1,505 @@
+}
+
+errno_t parse_token_result(struct devicecode_ctx *dc_ctx,
+                           char **error_description)
+{
+    json_t *tmp = NULL;
+    json_error_t json_error;
+
+    *error_description = NULL;
+    dc_ctx->result = json_loads(dc_ctx->http_data, 0, &json_error);
+    if (dc_ctx->result == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Failed to parse json data on line [%d]: [%s].\n",
+              json_error.line, json_error.text);
+        return EINVAL;
+    }
+
+    tmp = json_object_get(dc_ctx->result, "error");
+    if (json_is_string(tmp)) {
+        if (strcmp(json_string_value(tmp), "authorization_pending") == 0) {
+            json_decref(dc_ctx->result);
+            return EAGAIN;
+        } else {
+            *error_description = get_json_string(dc_ctx, dc_ctx->result,
+                                                 "error_description");
+            DEBUG(SSSDBG_OP_FAILURE, "Token request failed with [%s][%s].\n",
+                                     json_string_value(tmp),
+                                     *error_description);
+            json_decref(dc_ctx->result);
+            return EIO;
+        }
+    }
+
+    /* Looks like we got the tokens */
+    dc_ctx->access_token = json_object_get(dc_ctx->result, "access_token");
+    dc_ctx->access_token_str = get_json_string(dc_ctx, dc_ctx->result,
+                                               "access_token");
+    dc_ctx->id_token = json_object_get(dc_ctx->result, "id_token");
+    dc_ctx->id_token_str = get_json_string(dc_ctx, dc_ctx->result, "id_token");
+
+    return EOK;
+}
+
+static const char *get_id_string(TALLOC_CTX *mem_ctx, json_t *id_object)
+{
+    switch (json_typeof(id_object)) {
+    case JSON_STRING:
+        return json_string_value(id_object);
+        break;
+    case JSON_INTEGER:
+        return talloc_asprintf(mem_ctx, "%" JSON_INTEGER_FORMAT,
+                                        json_integer_value(id_object));
+        break;
+    default:
+        DEBUG(SSSDBG_CRIT_FAILURE,
+              "Unexpected user identifier type.\n");
+    }
+
+    return NULL;
+}
+
+const char *get_user_identifier(TALLOC_CTX *mem_ctx, json_t *userinfo,
+                                const char *user_identifier_attr)
+{
+    json_t *id_object = NULL;
+    const char *user_identifier = NULL;
+
+    if (user_identifier_attr != NULL) {
+        id_object = json_object_get(userinfo, user_identifier_attr);
+        if (id_object != NULL) {
+            user_identifier = get_id_string(mem_ctx, id_object);
+            if (user_identifier == NULL) {
+                DEBUG(SSSDBG_OP_FAILURE,
+                      "Failed to get user identifier string.\n");
+            }
+        } else {
+            DEBUG(SSSDBG_CRIT_FAILURE,
+                  "Failed to read attribute [%s] from userinfo data.\n",
+                  user_identifier_attr);
+        }
+    } else {
+        user_identifier = json_string_value(json_object_get(userinfo, "sub"));
```

Comment:
"sub" can be only string, as opposite to other options?

Reply:
yes, see e.g. https://openid.net/specs/openid-connect-core-1_0.html#Claims .